### ISSUE TITLE
GSoC: Compiling time optimizations to do

### ISSUE THREAD

**Issue by t13m** [Theano/Theano#2971]
This is a todo list of optimizations to make theano compile faster. Any idea and discussion is welcome! Thanks.
- [x] Remove unnecessary gpu_from_host(host_from_gpu(x)) ([ref](https://groups.google.com/d/msg/theano-dev/BiyppZg9ykA/IXpjNsCN7x0J))
- [x] Modify EquilibriumOptimizer to do less pass over the graph ([ref](https://groups.google.com/d/msg/theano-dev/BiyppZg9ykA/iPx7qLhwVm0J))
- [x] Make local_dimshuffle_list lift through many elemwise at once.
- [ ] Speed up inplace elemwise optimizer (issue #2962)
  - [x] Make some function inline
- [ ] This need discussion with other dev before being done. Make var.dimshuffle and var.reshape return the GPU op directly. This will remove work from the GPU. For this, go in the file cuda/var.py and create a new gpu_operator class that inherit from _operator and make CudaNdarrayVariable and CudaNdarrayConstant inherit from it, but not CudaNdarraySharedVariable.
- [x] Make local_fill_sink more efficient ([ref](https://groups.google.com/d/msg/theano-dev/BiyppZg9ykA/iPx7qLhwVm0J)) (not merged yet)
- [x] Merge nodes with input of assert node.
- [x] [Debug flag that print all rejected optimization](https://github.com/Theano/Theano/issues/2616)
- [x] Remove `merge3` (can't be done, it is needed)

later
- [ ] Find where the time spend during theano.function outside of linker and optimizer.

**Comment by nouiz** [comment]
@lamblin @abergeron what do you think of the 5th point that I added? I still want the graph to be build on the GPU, so I don't want to change what is done on shared variable. But if the user or during optimization we do method directly on a gpu variable, I think we could return directly the gpu op.

To test this and make sure it do not cause unforeseen problem, I would test it with dimshuffle and reshape.

**Comment by lamblin** [comment]
Do you mean that you "still want the graph to be built on the **CPU**" instead?
If so, that makes sense, although I'm not sure how much work is actually being done by the optimizer in that step.

**Comment by nouiz** [comment]
I mean, currently we var.dimshuffle will always create a dimshuffle on the
CPU. Even if var is a CudaNdarrayVariable.

What I want is that var.dimshuffle is a GpuDimShuffle when var is a
CudaNdarrayVariable.

This make optimization that build some graph like that do the stuff
directly on the GPU. So currently, it request other opt to move the
dimshuffle to the GPU. With the proposed change, this won't be needed. So
less change will be optimization application will be requested to move
everything to the GPU.

**Comment by lamblin** [comment]
I understand, but my questions are:
- why not do it for shared variables?
- how much time do you expect it to save during the optimization phase?

**Comment by nouiz** [comment]
Just to tell that I merged the 1st optimization. Can you have a first version of the second one by Wednesday?

**Comment by t13m** [comment]
Yes, I'm on it.

**Comment by nouiz** [comment]
For the next opt that was planned (locak_fill_sink), I think you didn't started the one that was there. Its logic is harder to understand and conter intuive, so I set it to later. You will have more experience and that will help.

So read again the list. The next one would be a new one: local_dimshuffle_list

Mostly, if you have this graph:

```
v=theano.tensor.vector()
m=theano.tensor.matrix()
out=(m+v)*m
```

then the dimshuffle on v will be lifted 2 times. This currently ask for 2 application of the opt local_dimshuffle_list. As this opt is applied very frequently 42s in canonizer and 39s in specialize, making this opt to move the dimshuffle over the 2 elemwise at the same time would speed up things quite a bit.

I already started that opt some time ago, but didn't had time to finish it. Here is the diff I had in case it help you start:

```
diff --git a/theano/tensor/opt.py b/theano/tensor/opt.py
index 438796b..dd24d58 100644
--- a/theano/tensor/opt.py
+++ b/theano/tensor/opt.py
@@ -439,6 +440,18 @@ def local_0_dot_x(node):
 ######################
 
 
+def apply_local_dimshuffle_lift(var):
+    return var
+    # lift recursively
+
+    if not var.owner:
+        return var
+    new = local_dimshuffle_lift.transform(var.owner)
+    if new:
+        return new[0]
+    return var
+
+
 @gof.local_optimizer([DimShuffle])
 def local_dimshuffle_lift(node):
     """
@@ -448,6 +461,7 @@ def local_dimshuffle_lift(node):
 
     DimShuffle(Elemwise(x, y)) => Elemwise(DimShuffle(x), DimShuffle(y))
     DimShuffle(DimShuffle(x)) => DimShuffle(x)
+    DimShuffle{0,1,...}(x) => x (when the dimshuffle do nothing)
 
     After this transform, clusters of Elemwise operations are
     void of DimShuffle operations.
@@ -460,23 +474,27 @@ def local_dimshuffle_lift(node):
     inode = input.owner
     if inode and isinstance(inode.op, Elemwise) and (len(input.clients) == 1):
         # Don't use make_node to have tag.test_value set.
-        ret = inode.op(*[op.__class__(inp.type.broadcastable,
-                                      op.new_order,
-                                      op.inplace)(inp) for inp in
-                         inode.inputs], **dict(return_list=True))
+        new_inputs = []
+        for inp in inode.inputs:
+            new_inp = op.__class__(inp.type.broadcastable,
+                                   op.new_order,
+                                   op.inplace)(inp)
+            new_inputs.append(apply_local_dimshuffle_lift(new_inp))
+        ret = inode.op(*new_inputs, **dict(return_list=True))
         return ret
     if inode and isinstance(inode.op, DimShuffle):
         new_order = [x == 'x' and 'x' or inode.op.new_order[x] for x in
                      op.new_order]
         inplace = op.inplace and inode.op.inplace
         iinput = inode.inputs[0]
+        # remove useless dimshuffle
         if new_order == range(len(new_order)) and (len(new_order) ==
                                                    iinput.type.ndim):
             return [iinput]
         else:
             ret = op.__class__(iinput.type.broadcastable, new_order,
-                               inplace)(iinput, **dict(return_list=True))
-            return ret
+                               inplace)(iinput)
+            return [apply_local_dimshuffle_lift(ret)]
 
 
 @register_canonicalize
```

Mostly, what it do it apply it once, then try to apply it again before returning something.

To test it, before profiling you can do something like this on the graph example above
```
theano.printing.debugprint(local_dimshuffle_list.transform(out.owner))

Then it should do the same as with the full compilation of the old version.

```

**Comment by t13m** [comment]
Thanks.

I don't fully understand your example. 
```
v = T.vector()
m = T.matrix()
out = (m + v) * m
```
generates following graph
```
Elemwise{mul,no_inplace} [@A] ''   
 |Elemwise{add,no_inplace} [@B] ''   
 | |<TensorType(float64, matrix)> [@C]
 | |DimShuffle{x,0} [@D] ''   
 |   |<TensorType(float64, vector)> [@E]
 |<TensorType(float64, matrix)> [@C]
```

The local_dimshuffle_lift makes changes like this:
```
DimShuffle(Elemwise(x, y)) => Elemwise(DimShuffle(x), DimShuffle(y))
DimShuffle(DimShuffle(x)) => DimShuffle(x)
```

How can it applied to such a graph with no DimShuffle(Elemwise()) or DimShuffle(Dimshuffle()) existing?

**Comment by nouiz** [comment]
This should be an example:

import theano
import theano.tensor as T

v = T.vector()
m = T.matrix()
out = (v+m).T
theano.config.optimizer_verbose=True
f=theano.function([v,m], out)
local_dimshuffle_lift DimShuffle{1,0}.0 Elemwise{add,no_inplace}.0
local_dimshuffle_lift DimShuffle{1,0}.0 DimShuffle{0,x}.0
dimshuffle_as_view DimShuffle{1,0}.0 InplaceDimShuffle{1,0}.0
dimshuffle_as_view DimShuffle{0,x}.0 InplaceDimShuffle{0,x}.0
inplace_elemwise_optimizer Elemwise{add,no_inplace}.0 Elemwise{Add}[(0,
1)].0


This show that the dimshffle opt was applied 2 times. The task is to have
the 2 optimization being done at once

Fred

**Comment by nouiz** [comment]
@lamblin about your question, if we do it for shared var, the user will build the graph on the GPU. This will probably cause other problem elsewhere in Theano. I prefer not to open that now without having time to make sure it don't cause problems.

**Comment by nouiz** [comment]
@t13m where are you on the new opt?

**Comment by t13m** [comment]
I made PR here https://github.com/Theano/Theano/pull/3014 . After testing I thought @nouiz 's patch worked well, so not sure what to do with it next.

**Comment by t13m** [comment]
Here ( https://github.com/t13m/Theano/tree/speed_up_inplace_elemwise_optimizer ) I roughly implement the algorithm described in issue #2962 . Didn't make a PR because it is untested.

**Comment by nouiz** [comment]
I added in the description a link to another ticket "debug flag..." It isn't hard, so if you end up being blocked on your current optimization, you could work on it.

**Comment by nouiz** [comment]
I'm not sure if you need more detail for the next step or not. But after the current PR, I think you should start work on the point related to local_fill_sink. Here is the snipet of doc about it in that link:
```
        7 - 4.183s 22 (0.332s in global opts, 0.321s io_toposort) - 14646 nodes - ('local_fill_sink', 12) ('local_fill_to_alloc', 7) ('constant_folding', 2) ('MergeOptimizer', 1)
        8 - 3.941s 15 (0.155s in global opts, 0.321s io_toposort) - 14641 nodes - ('local_fill_sink', 10) ('local_fill_to_alloc', 4) ('MergeOptimizer', 1)
        9 - 3.712s 11 (0.000s in global opts, 0.318s io_toposort) - 14639 nodes - ('local_fill_sink', 6) ('local_fill_to_alloc', 5)
       10 - 3.777s 6 (0.000s in global opts, 0.314s io_toposort) - 14635 nodes - ('local_fill_to_alloc', 4) ('local_fill_sink', 2)
       11 - 3.710s 2 (0.000s in global opts, 0.335s io_toposort) - 14631 nodes - ('local_fill_to_alloc', 1) ('local_fill_sink', 1)
       12 - 3.688s 1 (0.000s in global opts, 0.315s io_toposort) - 14630 nodes - ('local_fill_to_alloc', 1)

Could we change it to call itself recursively to do all of those iteration at the same time? To do this, it isn't trivial. At the end of the optimization, we need to check the clients of nodes, check if they are Elemwise and if so try this opt again on it with the replacement we generated. Then we need the local optimizer to return a dict with element. The key is the final node to replace in the graph and the value is its replacement. (this isn,t documented)
```

I think there is a few ways that this could be speed up. But it need some investigation.

1) First, manually check if there is case where the opt local_fill_sink could be applied recursively. But contrary to the opt local_dimshuffle_lift, the multiple apply will be done on the clients of the new node.
2) If you see that this could use useful, then start to implement it. As I tried to explain in the doc above, this will be an local optimizer, but will use a different interface of the local optimizer. What you will return is a dict, where the key are Apply node that you want to replace (in this case, I think there will be only one) and the value is a list of variable that you would return if you applied a local optimizer to that node.

3) Try if we could apply the  local_fill_to_alloc opt just after the canonicalize phase. I'm not sure this is the right way of doing it. I think we should apply all local_fill_sink first, then local_fill_to_alloc. But maybe local_fill_to_alloc need to be in the canonizer phase. As I'm not sure this is needed and the simplest solution would be to put it after, try this first.

If you have any ideas/comment for this case, do not hesitate to tell it. There is nothing sure about what the solution is.

**Comment by nouiz** [comment]
For 3), this should speed up compitation easily, but this could break Theano tests. So try to run all tests, including GPU tests. This could uncover problem by doing 3). Maybe there isn't, but maybe there is problems in doing that.

**Comment by t13m** [comment]
I think the local_fill_to_alloc should be applied after local_fill_sink, is optimizers in Canonizer optdb applied in an order? Anyway, I'm trying to implement the recursion first :)

**Comment by nouiz** [comment]
(The lower case/upper case are important). I'll try to make clear how this
work. But I'm sure I'll repeat many think you know. But as I'm not sure of
the questions, I'll give more information then needed. If that don't answer
your question, try to write it again.

There is at the top level, a sequential optimizer that have the name optdb.
optdb is also the name of a module that define different optimizer class.

in that top level seq optimizer, there is one optimizer called canonizer
that is an equilibrium. In that canonizer equilibrium, there is an instance
of a class called Canonizer. It have 2 instances: mul_canonizer and
add_canonizer. They are there to optimize elemwise like x/x=>1 and x-x=>1.
This is a local optimizer.

The equilibrium optimizer will run all local optimizer in order of them
being registered. So that mean currently you have this order (the line
number in my local file tensor/opt.py)

local_fill_to_alloc 1282
local_fill_sink 3404
local_mul_canonizer 3869
local_add_canonizer 4789

**Comment by t13m** [comment]
Thanks for explanation, it seems much clearer to me now.
By "canonizer equilibrium" do you mean the one named "canonicalize" registered here?
```
optdb.register('canonicalize', gof.EquilibriumDB(), 1, 'fast_run', 'fast_compile')
```

**Comment by nouiz** [comment]
Yes, that one.

**Comment by t13m** [comment]
Hi, @lamblin @nouiz , I want to begin the next optimization (as the 5th item listed), to make var.dimshuffle return GPU op. I think I'll firstly try the way described (creating new op class).

**Comment by nouiz** [comment]
closing as what is explicit here is done and there is other issue about the rest lest for speed up.
